Content.Server/Storage/EntitySystems/StorageSystem.cs

```diff
@@ -187,6 +191,234 @@ private void AddTransferVerbs(EntityUid uid, ServerStorageComponent component, G
+                    || !HasComp<SharedItemComponent>(target))
+                    return;
+
+                if (TryComp<TransformComponent>(uid, out var transformOwner) && TryComp<TransformComponent>(target, out var transformEnt))
+                {
+                    var position = EntityCoordinates.FromMap(
+                    transformOwner.Parent?.Owner ?? uid,
+                    transformEnt.MapPosition);
+                    if (PlayerInsertEntityInWorld(uid, eventArgs.User, target, storageComp))
+                    {
+                        RaiseNetworkEvent(new AnimateInsertingEntitiesEvent(uid,
+                            new List<EntityUid> { target },
+                            new List<EntityCoordinates> { position }));
+                    }
+                }
+            }
+            return;
+        }
+
+        private void OnDestroy(EntityUid uid, ServerStorageComponent storageComp, DestructionEventArgs args)
+        {
+            var storedEntities = storageComp.StoredEntities?.ToList();
+
+            if (storedEntities == null)
+                return;
+
+            foreach (var entity in storedEntities)
+            {
+                RemoveAndDrop(uid, entity, storageComp);
+            }
+        }
+
+        private void OnRemoveItemMessage(EntityUid uid, ServerStorageComponent storageComp, StorageRemoveItemMessage args)
+        {
+            if (args.Session.AttachedEntity == null)
+                return;
+
+            HandleRemoveEntity(uid, args.Session.AttachedEntity.Value, args.InteractedItemUID, storageComp);
+        }
+
+        private void OnInsertItemMessage(EntityUid uid, ServerStorageComponent storageComp, StorageInsertItemMessage args)
+        {
+            if (args.Session.AttachedEntity == null)
+                return;
+
+            PlayerInsertHeldEntity(uid, args.Session.AttachedEntity.Value, storageComp);
+        }
+
+        private void OnBoundUIOpen(EntityUid uid, ServerStorageComponent storageComp, BoundUIOpenedEvent args)
+        {
+            if (!storageComp.IsOpen)
+            {
+                storageComp.IsOpen = true;
+                UpdateStorageVisualization(uid, storageComp);
+            }
+        }
+
+        private void OnBoundUIClosed(EntityUid uid, ServerStorageComponent storageComp, BoundUIClosedEvent args)
+        {
+            if (TryComp<ActorComponent>(args.Session.AttachedEntity, out var actor) && actor?.PlayerSession != null)
+                CloseNestedInterfaces(uid, actor.PlayerSession, storageComp);
+
+            // If UI is closed for everyone
+            if (!_uiSystem.IsUiOpen(uid, args.UiKey))
+            {
+                storageComp.IsOpen = false;
+                UpdateStorageVisualization(uid, storageComp);
+
+                if (storageComp.StorageCloseSound is not null)
+                    SoundSystem.Play(Filter.Pvs(uid), storageComp.StorageCloseSound.GetSound(), uid, AudioParams.Default);
```

Comment:
Should pass in EntityManager to filters (they're kinda jank rn and should really be its own system).